File: src/p0497/java/solution.java
import java.util.Random;

class Solution {

    private int[][] rects;
    private int[]   areas;
    private int     total;
    private Random  random;

    public Solution(int[][] rects) {
        this.rects = rects;
        areas = new int[rects.length];
        total = 0;
        for (int i = 0; i < rects.length; i++) {
            areas[i] = total;
            total += (rects[i][2] - rects[i][0] + 1) * (rects[i][3] - rects[i][1] + 1);
        }
        random = new Random();
    }

    public int[] pick() {
        int area = random.nextInt(total);
        int upper = 0, lower = areas.length - 1;
        while (upper <= lower) {
            int mid = (upper + lower) / 2;
            if (areas[mid] <= area) {
                upper = mid + 1;
            } else {
                lower = mid - 1;
            }
        }
        area -= areas[lower];
        int dx = rects[lower][2] - rects[lower][0] + 1;
        int dy = rects[lower][3] - rects[lower][1] + 1;
        int x = rects[lower][0] + area % dx;
        int y = rects[lower][1] + area / dx;
        return new int[]{x, y};
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(rects);
 * int[] param_1 = obj.pick();
 */

File: src/p0497/cpp/solution.cpp
#include <iostream>
#include <random>
#include <vector>
#include <algorithm>

using namespace std;


class Solution {
public:
    Solution(vector<vector<int>> rects) : rects(rects) {
        total_area = 0;
        for (auto &r : rects) {
            total_area += (r[2] - r[0] + 1) * (r[3] - r[1] + 1);
            areas.push_back(total_area);
        }
        mt = mt19937(rd());
        dist = uniform_int_distribution<int>(0, total_area - 1);
    }

    vector<int> pick() {
        auto area = dist(mt);
        auto it = upper_bound(areas.begin(), areas.end(), area);
        area -= (it == areas.begin() ? 0 : *(it - 1));
        auto r = rects[it - areas.begin()];
        auto dx = r[2] - r[0] + 1, dy = r[3] - r[1] + 1;
        return {r[0] + area % dx, r[1] + area / dx};
    }

private:
    vector<vector<int>> rects;
    vector<int> areas;
    int total_area;
    random_device rd;
    mt19937 mt;
    uniform_int_distribution<int> dist;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(rects);
 * vector<int> param_1 = obj.pick();
 */
